packages/editor/src/components/post-actions/set-as-homepage.js

```diff
@@ -0,0 +1,216 @@
+				page_on_front: item.id,
+				show_on_front: 'page',
+			} );
+
+			createSuccessNotice( __( 'Homepage updated' ), {
+				type: 'snackbar',
+			} );
+		} catch ( error ) {
+			const typedError = error;
+			const errorMessage =
+				typedError.message && typedError.code !== 'unknown_error'
+					? typedError.message
+					: __( 'An error occurred while setting the homepage' );
+			createErrorNotice( errorMessage, { type: 'snackbar' } );
+		}
+	}
+
+	const renderModalBody = () => {
+		if ( 'posts' === showOnFront ) {
+			return (
+				<>
+					<Text>
+						{ sprintf(
+							// translators: %s: title of the page to be set as the homepage.
+							__(
+								'Set "%s" as the site homepage? This will replace the current homepage which is set to display latest posts.'
+							),
+							pageTitle
+						) }
+					</Text>
+				</>
+			);
+		}
+
+		if ( isPageDraft ) {
+			return (
+				<>
+					<Text>
+						{ sprintf(
+							// translators: %s: title of the page to be set as the homepage.
+							__(
+								'"%s" is a draft and will be published automatically if set as the homepage. Publish page and set as the site homepage?'
+							),
+							pageTitle
+						) }
+					</Text>
+				</>
+			);
+		}
+
+		const modalTranslatedString =
+			// translators: %1$s: title of page to be set as the home page. %2$s: title of the current home page.
+			__(
+				'Set "%1$s" as the site homepage? This will replace the current homepage: "%2$s"'
+			);
+
+		return (
+			<Text>
+				{ sprintf(
+					modalTranslatedString,
+					pageTitle,
+					currentHomePageTitle
+				) }{ ' ' }
+			</Text>
+		);
+	};
+
+	// translators: Button label to confirm setting the specified page as the homepage.
+	let modalButtonLabel = __( 'Set homepage' );
+	if ( isPageDraft ) {
+		modalButtonLabel = __( 'Publish and set homepage' );
+	}
+
+	return (
+		<form onSubmit={ onSetPageAsHomepage }>
+			<VStack spacing="5">
+				{ renderModalBody() }
+				<HStack justify="right">
+					<Button
+						__next40pxDefaultSize
+						variant="tertiary"
+						onClick={ () => {
+							closeModal?.();
+						} }
+						disabled={ isSavingSiteSettings }
+						accessibleWhenDisabled
+					>
+						{ __( 'Cancel' ) }
+					</Button>
+					<Button
+						__next40pxDefaultSize
+						variant="primary"
+						type="submit"
+						disabled={ isSavingSiteSettings }
+						accessibleWhenDisabled
+					>
+						{ modalButtonLabel }
+					</Button>
+				</HStack>
+			</VStack>
+		</form>
+	);
+};
+
+export const useSetAsHomepageAction = () => {
+	const { pageOnFront, pageForPosts } = useSelect( ( select ) => {
+		const { getEntityRecord } = select( coreStore );
+		const siteSettings = getEntityRecord( 'root', 'site' );
+		return {
+			pageOnFront: siteSettings?.page_on_front,
+			pageForPosts: siteSettings?.page_for_posts,
+		};
+	} );
+
+	return useMemo(
+		() => ( {
+			id: 'set-as-homepage',
+			label: __( 'Set as homepage' ),
+			isEligible( post ) {
+				if ( post.status === 'trash' ) {
```

Comment:
```suggestion
				if ( post.status === 'trash' ) {
```

I'd like to know which post statuses of the following [defaults](https://wordpress.org/documentation/article/post-status/#default-statuses) should _not_ be allowed to set as the homepage. 

- Publish
- Future
- Draft
- Pending
- Private
- Trash
- Auto-Draft

Note also we cannot use a whitelist because it's possible to add custom statuses.

Do we 

- restrict the display of the action?
- allow all status (except trssh) and add more cases in the modal to explain the consequences of moving each status (including a fallback for unknown custom statuses) to `publish`?

I tested with `Pending` status and the messaging and wording wasn't completely clear.

Reply:
Good call. I think I'd prefer to restrict the display of the action to only `publish` or `draft` statuses.

Pinging @jameskoster in case you have any thoughts here.

Reply:
Starting simple makes sense to me. I forget if we already discussed this but we might consider restricting only to published pages for now, same as wp-admin. This might help with the performance issue noted above?

Allowing draft pages (including the publish flow) could be something to explore later.

Reply:
> Allowing draft pages (including the publish flow) could be something to explore later.

I agree that this is a good follow-up step, and then we can concentrate on getting the most simple form of the action merged in this PR.

It felt like an easy addition when I first started working on it, but there are a few gotchas with handling draft pages:

- We need to inform users that the page status is changing from draft to publish. This PR currently handles that in the action modal wording, but I'm not convinced this is the best way to handle this.
- Updating a page's status doesn't seem to be easily observable in the global state, which is part of the performance issue noted above. I believe this is an existing issue with the `isSavingEntityRecord` function, so it'd be nice to address this in a separate PR.